confdb/confdb.go

```diff
@@ -470,42 +523,105 @@ type validationOptions struct {
 //   - non-placeholder subkeys are made up of lowercase alphanumeric ASCII characters,
 //     optionally with dashes between alphanumeric characters (e.g., "a-b-c")
 //   - placeholder subkeys are composed of non-placeholder subkeys wrapped in curly brackets
-func validateViewDottedPath(path string, opts *validationOptions) (err error) {
+func validateViewDottedPath(path string, opts *validationOptions) error {
 	if opts == nil {
 		opts = &validationOptions{}
 	}
 
-	subkeys := strings.Split(path, ".")
+	subkeys, err := splitViewPath(path)
+	if err != nil {
+		return err
+	}
+
 	for _, subkey := range subkeys {
-		if subkey == "" {
-			return errors.New("cannot have empty subkeys")
-		}
+		// straight literal accesses, without placeholders (e.g., foo.bar, foo[1])
+		isLiteral := validSubkey.MatchString(subkey) || validIndexSubkey.MatchString(subkey)
+		// placeholder subkeys (e.g., foo.{bar}, foo[{n}])
+		isPlaceholder := validPlaceholder.MatchString(subkey) || validIndexPlaceholder.MatchString(subkey)
 
-		if !validSubkey.MatchString(subkey) && (!opts.allowPlaceholder || !validPlaceholder.MatchString(subkey)) {
+		if !isLiteral && (!opts.allowPlaceholder || !isPlaceholder) {
 			return fmt.Errorf("invalid subkey %q", subkey)
 		}
 	}
 
 	return nil
 }
 
-// getPlaceholders returns the set of placeholders in the string or nil, if
-// there is none.
-func getPlaceholders(viewStr string) map[string]bool {
-	var placeholders map[string]bool
+func splitViewPath(path string) ([]string, error) {
```

Comment:
wouldn't it make sense to return something a bit more complex than string here, like an interface that can deal with index vs key more nicely?

Reply:
ah so just moving the isIndex(), isPlaceholder(), etc methods into a type. I can do that, sure. The diff will be pretty large though since lots of code depends on these paths

Reply:
yes, but it also mean they can return a constant if we use different types and we do the recognizing when we split. Also you can add extra methods as well, if they have more than a couple usage places